flang/lib/Optimizer/OpenMP/DoConcurrentConversion.cpp

```diff
@@ -548,19 +357,16 @@ class DoConcurrentConversionPass
       return;
     }
 
-    llvm::DenseSet<fir::DoLoopOp> concurrentLoopsToSkip;
+    llvm::DenseSet<fir::DoConcurrentOp> concurrentLoopsToSkip;
```

Comment:
The correctness of this depends upon the traversal order used by `applyFullConversion`. I'm not certain that is guaranteed. Also modifying this (conceptually) global list is harder to follow. Now that nesting is much simpler, could we instead now skip any `fir::DoConcurrentOp` which has an ancestor which is a `fir::DoConcurrentOp`?

Reply:
> I'm not certain that is guaranteed.

ops are walked [in pre-order](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Transforms/Utils/DialectConversion.cpp#L2679). So the parent op is guaranteed to be traversed before the nested ones.

> could we instead now skip any fir::DoConcurrentOp which has an ancestor which is a fir::DoConcurrentOp?

The problem is that we clone region of the parent `do concurrent` op to the newly created set of OMP ops including any nested `do concurrent` loops. Therefore, the condition of checking for a `do concurrent` ansector will not apply anymore.

I also do not like the gloabl list that much tbh. I am trying to think of a better solution.